atc/pipeline.go

```diff
@@ -39,37 +37,19 @@ type RenameRequest struct {
 
 type InstanceVars map[string]interface{}
 
-func (iv InstanceVars) Flatten() (DotNotation, error) {
-	return flatten("", 0, (map[string]interface{})(iv))
-}
-
 func (iv InstanceVars) String() string {
-	m, _ := iv.Flatten()
-	keys := make([]string, 0, len(m))
-	for k := range m {
-		keys = append(keys, k)
-	}
-	sort.Strings(keys)
-
-	var instanceVars []string
-	for _, k := range keys {
-		bytes, _ := yaml.Marshal(m[k])
-		bytes = []byte(strings.TrimRight(string(bytes), "\n"))
-		instanceVars = append(instanceVars, fmt.Sprintf("%s:%s", k, bytes))
-	}
-	return strings.Join(instanceVars, ",")
-}
-
-type DotNotation map[string]interface{}
-
-func (d DotNotation) Expand() (InstanceVars, error) {
-	var out interface{}
-	for key, value := range d {
-		if err := expand(&out, strings.Split(key, ".")[:], value); err != nil {
-			return nil, err
-		}
+	kvPairs := vars.StaticVariables(iv).Flatten()
+	sort.Slice(kvPairs, func(i, j int) bool {
+		return kvPairs[i].Ref.String() < kvPairs[j].Ref.String()
+	})
+
+	var parts []string
+	for _, kvPair := range kvPairs {
+		rawVal, _ := yaml.Marshal(kvPair.Value)
+		val := strings.TrimRight(string(rawVal), "\n")
```

Comment:
If the `Value` contains a comma, the entire value will need to be quoted or else it can be confused with the separator for another instance var. 

I.e. `fly -t dev up -p p/foo:bar,baz` breaks while  `fly -t dev up -p "p/foo:\"bar,baz\""` is the correct form

Reply:
Ah, good catch. I also realized that marshaling values as YAML doesn't work for so well for lists, since it'll do the

```
- val1
- val2
```

syntax. Going to switch this to just marshaling as JSON, and we can explicitly unquote the value if it's a string and it doesn't contain any special characters